### PR TITLE
Replace socket select() with poll()

### LINKED ISSUE #6264: Replace using of `select()` with `poll()`
**Task (REQUIRED):**
image

Sometimes SDKs use more than 1024 FD, if the engine a game use API that uses `select()`  under the hood at the same time, than the game crashes.

The engine uses `select()`  in 6 files:
`qsa.c`   - openAl/Alc
`wsocket.c`  - luasocket
`usocket.c`  - luasocket
`net_sockets.c`  - mdedtls
`sslsocket.c` - dlib
`socket.cpp`  - dlib

`select()` ([doc](https://man7.org/linux/man-pages/man2/select.2.html)) should be replaced with `poll()` ([doc](https://man7.org/linux/man-pages/man2/poll.2.html)) to avoid this crash.

**Expected outcome (REQUIRED):**

No more crashes in such cases.
A couple of examples:

```
...
#00  pc 0000000000000668  [vdso] (__kernel_rt_sigreturn)
  #00  pc 000000000004e408  /apex/com.android.runtime/lib64/bionic/libc.so (abort+160)
  #00  pc 0000000000076c00  /apex/com.android.runtime/lib64/bionic/libc.so (__fortify_fatal(char const*, ...)+124)
  #00  pc 0000000000076b1c  /apex/com.android.runtime/lib64/bionic/libc.so (__FD_SET_chk+120)
  #00  pc 00000000001f4d08 (dmConnectionPool::ConnectSocket(dmConnectionPool::ConnectionPool*, dmSocket::Address, unsigned short, int, dmConnectionPool::Connection*, dmSocket::Result*)) (SourceCode: /home/runner/work/defold/defold/engine/dlib/build/../src/dlib/connection_pool.cpp:273)
...
```

One more crash:
https://github.com/defold/extension-poco/blob/master/poco/lua/defold-poco.lua#L146

### PR DESCRIPTION
**Pull request by britzl**
Fixes #6264

### CODE REVIEW

File: engine/dlib/src/dlib/file_descriptor.cpp [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+#include <stdint.h>
+#include <assert.h>
+#if defined(_WIN32)
+#include <winsock2.h>
+#else
+#include <poll.h>
+#endif
+
+
+namespace dmFileDescriptor
```
Comment (britzl) [inline engine/dlib/src/dlib/file_descriptor.cpp:24]: I'm not entirely sure about the naming here. This new namespace currently only contains functionality to use poll() on one or more file descriptors. Could we move this somewhere else?

File: engine/dlib/src/dlib/file_descriptor.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        pollfd pfd;
+        pfd.fd = fd;
+        pfd.events = e;
+        poller->m_Pollfds.Push(pfd);
+        #endif
+    }
+
+    bool PollerHasEvent(Poller* poller, PollEvent event, int fd)
+    {
+        for (uint32_t i = 0; i < poller->m_Pollfds.Size(); ++i)
```
Comment (JCash) [inline engine/dlib/src/dlib/file_descriptor.cpp:119]: How large are these arrays?

File: engine/dlib/src/dmsdk/dlib/file_descriptor.h [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+     * @name Poller
+     */
+    struct Poller
+    {
+        #if defined(_WIN32)
+        dmArray<WSAPOLLFD> m_Pollfds;
+        #else
+        dmArray<pollfd> m_Pollfds;
+        #endif
+    };
```
Comment (JCash) [inline engine/dlib/src/dmsdk/dlib/file_descriptor.h:60]: This makes it really hard to make it platform independent for our platforms.
Use a typedef instead.

**Review by JCash — CHANGES_REQUESTED** [review]

File: engine/dlib/src/dlib/file_descriptor.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+#define DM_FILE_DESCRIPTOR_H
+
+#include <dmsdk/dlib/file_descriptor.h>
+
+/**
+ * File descriptor abstraction
+ */
+namespace dmFileDescriptor
+{
+}
```
[the remark this answers is not in the document]
Comment (britzl) [reply engine/dlib/src/dlib/file_descriptor.h]: Removed

File: engine/dlib/src/dlib/file_descriptor.cpp [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+#include "file_descriptor.h"
+#include "log.h"
+#include <stdint.h>
+#include <assert.h>
+#if defined(_WIN32)
+#include <winsock2.h>
+#else
+#include <poll.h>
+#endif
```
[the remark this answers is not in the document]
Comment (britzl) [reply engine/dlib/src/dlib/file_descriptor.cpp]: Good point. Added!

File: engine/dlib/src/dlib/file_descriptor.cpp [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+            case EVENT_READ: return POLLRDNORM;
+            case EVENT_WRITE: return POLLWRNORM;
+            default:
+            case EVENT_ERROR: return POLLRDBAND;
+        }
+        #else
+        switch (event)
+        {
+            case EVENT_READ: return POLLIN;
+            // case EVENT_READ: return POLLIN | POLLPRI;
```
[the remark this answers is not in the document]
Comment (britzl) [reply engine/dlib/src/dlib/file_descriptor.cpp]: Removed the commented code

File: engine/dlib/src/dlib/socket.cpp [hunk 7]
```diff
@@ -1100,42 +1101,50 @@ namespace dmSocket
...
+    dmFileDescriptor::PollEvent SelectorKindToPollEvent(SelectorKind selector_kind)
+    {
+        switch (selector_kind)
+        {
+            case SELECTOR_KIND_READ:
+                return dmFileDescriptor::EVENT_READ;
+            case SELECTOR_KIND_WRITE:
+                return dmFileDescriptor::EVENT_WRITE;
+            default:
+            case SELECTOR_KIND_EXCEPT:
```
[the remark this answers is not in the document]
Comment (britzl) [reply engine/dlib/src/dlib/socket.cpp]: Moved default and added assert()

File: engine/dlib/src/dlib/file_descriptor.cpp (hunk 2 above)
Comment (britzl) [reply engine/dlib/src/dlib/file_descriptor.cpp:63]: Quite often just a single file descriptor, sometimes a handful. It depends on where they're used.

File: engine/dlib/src/dmsdk/dlib/file_descriptor.h (hunk 3 above)
Comment (britzl) [reply engine/dlib/src/dmsdk/dlib/file_descriptor.h]: Yep, added!

**Review by JCash — APPROVED** [review]

**Pull request merged**